src/main/java/org/commcare/session/CommCareSession.java

```diff
@@ -949,14 +959,28 @@ public static CommCareSession restoreSessionFromStream(CommCarePlatform ccPlatfo
             stackFrames.push(lastElement);
         }
         restoredSession.setFrameStack(stackFrames);
-        restoredSession.syncState();
 
+        for (StackFrameStep step : restoredFrame.getSteps()) {
+            if (step.hasXmlInstance()) {
+                TreeElement root = (TreeElement)ExtUtil.read(inputStream, TreeElement.class, null);
+                step.setXmlInstance(ExternalDataInstance.buildFromRemote(
+                        step.getXmlInstance().getInstanceId(), root, step.getXmlInstance().useCaseTemplate()));
+
+            }
+        }
+
+        restoredSession.syncState();
         return restoredSession;
     }
 
     public void serializeSessionState(DataOutputStream outputStream) throws IOException {
         frame.writeExternal(outputStream);
         ExtUtil.write(outputStream, new ExtWrapList(frameStack));
+        for (StackFrameStep step : frame.getSteps()) {
```

Comment:
This might have potential impact on disk storage and Clayton suggested that we add some sort of a threshold check here before serializing the xml. I still need to add that into this PR.

Reply:
why do we serialise this here with the session instead as part of the SessionFrame?

Reply:
This serialization ideally belongs in [ExternalDataInstance](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L99), though because of [root](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L19) being an instance of  interface `AbstractTreeElement` it skips searializing it over there. Here I am assuming that the `root` is always an `TreeElement` to serialize it and thought since we are alreay handling some serialization here it would be the next best place to put this code here. Though this does look a bit sketchy and I would be open to ideas here. Also tagging @ctsims to get this thoughts on the same

Reply:
Another issue I am facing to implement the size thereshold is we don't seem to have a way to calculate the size of a `TreeElement` without serializing it and checking the byte size. So I am worried if we implement this we will need to serialize the instance xml twice ?

Reply:
@shubham1g5 

I think it's a bit dangerous to serialize this way directly for the reasons you mentioned. It's hard to gauge the size impact, and if the instance is serialized as the mechanism of persistence it makes it harder to set multiple paths for an instance to recover the instance data. 

My sense is that what we want is to extend the abstraction for **providing** the external input that gets turned into the data instance, and make it possible to either re-fetch data, or store it.

This is the entrypoint which I was thinking of:
https://github.com/dimagi/commcare-core/blob/f15e37c497fd6a1fc4511b8902dc13634fe607e6/src/main/java/org/commcare/session/RemoteQuerySessionManager.java#L144

If we could wrap and defer "Here's the source of an instance" we could serialize that in a way which can be retrieved. 

Then as part of events during the session lifecycle we could push the datum elements to go from "source" to "populated" and during that process determine whether to temporarily cache that response, permanently cache that response, or re-request the response at the appropriate point in the lifecycle.

Reply:
Sounds good to me. @shubham1g5 do you have a sense of how to go about doing that?

Reply:
@ctsims @snopoke I am trying adding a [remoteUrl](https://github.com/dimagi/commcare-core/blob/dr_form_query/src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java#L33) field in the `ExternalDataInstance` and retrieiving it later from cache/web while [initializing the FormSession](https://github.com/dimagi/formplayer/blob/dr_form_query/src/main/java/org/commcare/formplayer/session/FormSession.java#L228).

I am struggling with how to best handle failures on retrieving the instance data though one possible way can be to pass a flag to HQ that simply moves user one step back and we ask the user to select the form again so that we can run through the `QueryScreen` code again ? 

Also @ctsims, I am not able to find any obvious hooks into the `CommCareInstanceInitializer` to put this code in that makes it easeir for us to maintain the query cycle. Would definitely appreciate any tips on how to best organise this code.